File: tests/data_offloading.cpp
#include <simgrid/s4u.hpp>
#include <iostream>
#include <vector>

XBT_LOG_NEW_DEFAULT_CATEGORY(data_offloading, "Data Offloading Application");

namespace sg4 = simgrid::s4u;

/**
 * @brief Data offloading application
 * 
 * Simulates offloading decisions: process locally on Edge,
 * send to Fog, or send to Cloud based on load and resources.
 */

class SmartEdgeDevice {
    std::string fog_name;
    std::string cloud_name;
    double workload;
    int num_tasks;
    
public:
    SmartEdgeDevice(const std::string& fog, const std::string& cloud, double load, int tasks = 1)
        : fog_name(fog), cloud_name(cloud), workload(load), num_tasks(tasks) {}
    
    void operator()() {
        sg4::Host* this_host = sg4::this_actor::get_host();
        std::string result_mailbox_name = std::string(this_host->get_cname()) + "_result";
        
        XBT_INFO("Edge Device '%s' started (load: %.2f GFlops, tasks: %d)",
                 this_host->get_cname(), workload / 1e9, num_tasks);
        
        double device_capacity = this_host->get_speed();
        
        // Process multiple tasks
        for (int task_id = 0; task_id < num_tasks; task_id++) {
            XBT_INFO("--- Processing task %d/%d ---", task_id + 1, num_tasks);
            
            // Offloading decision based on capacity
            if (workload < device_capacity * 0.5) {
                // Local processing - low load
                XBT_INFO("Decision: LOCAL PROCESSING (low load)");
                double start_time = sg4::Engine::get_clock();
                sg4::this_actor::execute(workload);
                double end_time = sg4::Engine::get_clock();
                XBT_INFO("Local processing completed in %.2f seconds",
                         end_time - start_time);
                
            } else if (workload < device_capacity * 2.0) {
                // Offload to Fog - medium load
                XBT_INFO("Decision: OFFLOAD TO FOG (medium load)");
                
                auto* mbox = sg4::Mailbox::by_name(fog_name);
                
                // Create task payload with source info
                auto* task_data = new std::pair<double, std::string>(workload, result_mailbox_name);
                
                XBT_INFO("Sending task to Fog '%s' (%.2f MFlops)...",
                         fog_name.c_str(), workload / 1e6);
                mbox->put(task_data, workload / 1e3);  // Data size = workload/1000
                
                // Wait for result
                auto* result_mbox = sg4::Mailbox::by_name(result_mailbox_name);
                auto* result = result_mbox->get<std::string>();
                XBT_INFO("Result received from Fog: %s", result->c_str());
                delete result;
                
            } else {
                // Offload to Cloud - high load
                XBT_INFO("Decision: OFFLOAD TO CLOUD (high load)");
                
                auto* mbox = sg4::Mailbox::by_name(cloud_name);
                
                // Create task payload with source info
                auto* task_data = new std::pair<double, std::string>(workload, result_mailbox_name);
                
                XBT_INFO("Sending task to Cloud '%s' (%.2f MFlops)...",
                         cloud_name.c_str(), workload / 1e6);
                mbox->put(task_data, workload / 1e3);
                
                // Wait for result
                auto* result_mbox = sg4::Mailbox::by_name(result_mailbox_name);
                auto* result = result_mbox->get<std::string>();
                XBT_INFO("Result received from Cloud: %s", result->c_str());
                delete result;
            }
            
            // Small delay between tasks
            if (task_id < num_tasks - 1) {
                sg4::this_actor::sleep_for(0.5);
            }
        }
        
        XBT_INFO("All %d tasks completed successfully", num_tasks);
    }
};

class OffloadingServer {
    std::string type;
    int max_tasks;
    
public:
    OffloadingServer(const std::string& server_type, int max = -1) 
        : type(server_type), max_tasks(max) {}
    
    void operator()() {
        sg4::Host* this_host = sg4::this_actor::get_host();
        XBT_INFO("[%s] Server '%s' ready to receive tasks",
                 type.c_str(), this_host->get_cname());
        
        auto* mbox = sg4::Mailbox::by_name(this_host->get_cname());
        int tasks_processed = 0;
        
        // Server listening for tasks
        while (max_tasks < 0 || tasks_processed < max_tasks) {
            try {
                // Receive task with source mailbox info
                auto* task_data = mbox->get<std::pair<double, std::string>>(10.0);  // Timeout 10s
                
                double workload = task_data->first;
                std::string reply_to = task_data->second;
                
                XBT_INFO("[%s] Task received: %.2f MFlops (reply to: %s)", 
                         type.c_str(), workload / 1e6, reply_to.c_str());
                
                // Process task
                double start_time = sg4::Engine::get_clock();
                sg4::this_actor::execute(workload);
                double end_time = sg4::Engine::get_clock();
                
                XBT_INFO("[%s] Task processed in %.2f seconds",
                         type.c_str(), end_time - start_time);
                
                tasks_processed++;
                delete task_data;
                
                // Send result back to the requesting device
                auto* result_mbox = sg4::Mailbox::by_name(reply_to);
                auto* result = new std::string(
                    type + " processed task in " + 
                    std::to_string(end_time - start_time) + "s");
                
                result_mbox->put(result, 100);  // Small result message
                XBT_INFO("[%s] Result sent back to %s", type.c_str(), reply_to.c_str());
                
            } catch (const simgrid::TimeoutException&) {
                // Timeout - no more tasks
                XBT_INFO("[%s] No more tasks, finishing (processed %d tasks)", 
                         type.c_str(), tasks_processed);
                break;
            }
        }
        
        if (max_tasks > 0 && tasks_processed >= max_tasks) {
            XBT_INFO("[%s] Maximum tasks reached (%d), server finishing", 
                     type.c_str(), tasks_processed);
        }
        
        XBT_INFO("[%s] Server finished (total tasks: %d)", type.c_str(), tasks_processed);
    }
};

int main(int argc, char* argv[]) {
    sg4::Engine e(&argc, argv);
    
    if (argc < 2) {
        XBT_CRITICAL("Usage: %s <platform_file.xml> [num_tasks_per_device]", argv[0]);
        XBT_CRITICAL("Example: %s platforms/hybrid_platform.xml 3", argv[0]);
        XBT_CRITICAL("  - num_tasks_per_device: Number of tasks each edge device will send (default: 1)");
        return 1;
    }
    
    // Parse number of tasks per device (optional parameter)
    int num_tasks_per_device = 10;
    if (argc >= 3) {
        num_tasks_per_device = std::atoi(argv[2]);
        if (num_tasks_per_device < 1) {
            XBT_CRITICAL("num_tasks_per_device must be >= 1");
            return 1;
        }
    }
    
    e.load_platform(argv[1]);
    
    std::vector<sg4::Host*> hosts = e.get_all_hosts();
    
    if (hosts.size() < 3) {
        XBT_CRITICAL("At least 3 hosts required (Edge, Fog, Cloud)");
        return 1;
    }
    
    XBT_INFO("=== Data Offloading Application ===");
    XBT_INFO("Platform loaded with %zu hosts", hosts.size());
    XBT_INFO("Tasks per edge device: %d", num_tasks_per_device);
    
    // Classify hosts
    std::vector<sg4::Host*> edge_hosts, fog_hosts, cloud_hosts;
    
    for (auto* host : hosts) {
        std::string name = host->get_cname();
        if (name.find("edge") != std::string::npos) {
            edge_hosts.push_back(host);
        } else if (name.find("fog") != std::string::npos) {
            fog_hosts.push_back(host);
        } else if (name.find("cloud") != std::string::npos) {
            cloud_hosts.push_back(host);
        }
    }
    
    // Default distribution if no specific names
    if (edge_hosts.empty() && fog_hosts.empty() && cloud_hosts.empty()) {
        size_t third = hosts.size() / 3;
        edge_hosts.assign(hosts.begin(), hosts.begin() + third);
        fog_hosts.assign(hosts.begin() + third, hosts.begin() + 2 * third);
        cloud_hosts.assign(hosts.begin() + 2 * third, hosts.end());
    }
    
    if (edge_hosts.empty()) edge_hosts.push_back(hosts[0]);
    if (fog_hosts.empty()) fog_hosts.push_back(hosts[std::min(1UL, hosts.size()-1)]);
    if (cloud_hosts.empty()) cloud_hosts.push_back(hosts.back());
    
    XBT_INFO("Configuration:");
    XBT_INFO("  Edge devices: %zu", edge_hosts.size());
    XBT_INFO("  Fog nodes: %zu", fog_hosts.size());
    XBT_INFO("  Cloud servers: %zu", cloud_hosts.size());
    
    // Calculate expected total tasks for servers
    int expected_offloaded_tasks = 0;
    for (size_t i = 0; i < edge_hosts.size(); i++) {
        double workload = 1e9 * (0.5 + i * 0.5);
        double device_capacity = edge_hosts[i]->get_speed();
        // Only count tasks that will be offloaded
        if (workload >= device_capacity * 0.5) {
            expected_offloaded_tasks += num_tasks_per_device;
        }
    }
    
    XBT_INFO("  Expected offloaded tasks: %d", expected_offloaded_tasks);
    
    // Create Fog servers (with task limit based on expected load)
    for (auto* fog : fog_hosts) {
        fog->add_actor("fog_server", OffloadingServer("FOG", -1));  // -1 = unlimited
    }
    
    // Create Cloud servers (with task limit based on expected load)
    for (auto* cloud : cloud_hosts) {
        cloud->add_actor("cloud_server", OffloadingServer("CLOUD", -1));  // -1 = unlimited
    }
    
    // Create Edge devices with different loads
    double base_workload = 1e9;  // 1 GFlop base
    for (size_t i = 0; i < edge_hosts.size(); i++) {
        double workload = base_workload * (0.5 + i * 0.5);  // Increasing load
        
        std::string fog_name = fog_hosts[i % fog_hosts.size()]->get_cname();
        std::string cloud_name = cloud_hosts[i % cloud_hosts.size()]->get_cname();
        
        edge_hosts[i]->add_actor("smart_device_" + std::to_string(i),
                                 SmartEdgeDevice(fog_name, cloud_name, workload, num_tasks_per_device));
    }
    
    // Run simulation
    e.run();
    
    XBT_INFO("=== Simulation completed ===");
    XBT_INFO("Simulated time: %.2f seconds", sg4::Engine::get_clock());
    
    return 0;
}

File: src/python/tests/data_offloading.py
"""
data_offloading.py – Adaptive offloading: Edge, Fog, or Cloud based on load.

Mirrors tests/data_offloading.cpp.

Usage
-----
    python data_offloading.py <platform_file.xml>
"""
from __future__ import annotations

import sys
import os

_HERE = os.path.dirname(os.path.abspath(__file__))
_ROOT = os.path.join(_HERE, "..")
if _ROOT not in sys.path:
    sys.path.insert(0, _ROOT)

import simgrid


class SmartEdgeDevice:
    """
    Makes runtime offloading decisions:
    - Low workload  → process locally
    - Medium workload → offload to Fog
    - High workload → offload to Cloud
    """

    def __init__(
        self,
        fog_name: str,
        cloud_name: str,
        workload: float,
        num_tasks: int = 1,
    ) -> None:
        self._fog       = fog_name
        self._cloud     = cloud_name
        self._workload  = workload
        self._num_tasks = num_tasks

    def __call__(self) -> None:
        host = simgrid.this_actor.get_host()
        result_mbox_name = f"{host.name}_result"

        simgrid.this_actor.info(
            f"Edge Device '{host.name}' started "
            f"(workload: {self._workload / 1e9:.2f} GFlops, tasks: {self._num_tasks})"
        )

        capacity = host.speed

        for task_id in range(self._num_tasks):
            simgrid.this_actor.info(
                f"--- Processing task {task_id + 1}/{self._num_tasks} ---"
            )

            if self._workload < capacity * 0.5:
                # Local processing
                simgrid.this_actor.info("Decision: LOCAL PROCESSING (low load)")
                t0 = simgrid.Engine.clock
                simgrid.this_actor.execute(self._workload)
                t1 = simgrid.Engine.clock
                simgrid.this_actor.info(f"Local processing completed in {t1 - t0:.2f} s")

            elif self._workload < capacity * 2.0:
                # Offload to Fog
                simgrid.this_actor.info("Decision: OFFLOAD TO FOG (medium load)")
                mbox = simgrid.Mailbox.by_name(self._fog)
                simgrid.this_actor.info(
                    f"Sending task to Fog '{self._fog}' "
                    f"({self._workload / 1e6:.2f} MFlops)…"
                )
                mbox.put((self._workload, result_mbox_name), int(self._workload / 1e3))

                result = simgrid.Mailbox.by_name(result_mbox_name).get()
                simgrid.this_actor.info(f"Result received from Fog: {result}")

            else:
                # Offload to Cloud
                simgrid.this_actor.info("Decision: OFFLOAD TO CLOUD (high load)")
                mbox = simgrid.Mailbox.by_name(self._cloud)
                simgrid.this_actor.info(
                    f"Sending task to Cloud '{self._cloud}' "
                    f"({self._workload / 1e6:.2f} MFlops)…"
                )
                mbox.put((self._workload, result_mbox_name), int(self._workload / 1e3))

                result = simgrid.Mailbox.by_name(result_mbox_name).get()
                simgrid.this_actor.info(f"Result received from Cloud: {result}")

            if task_id < self._num_tasks - 1:
                simgrid.this_actor.sleep_for(0.5)

        simgrid.this_actor.info(f"All {self._num_tasks} tasks completed successfully")


class OffloadingServer:
    """
    Generic compute server (Fog or Cloud) that:
    1. Receives a (workload_flops, return_mailbox) tuple.
    2. Executes the workload.
    3. Sends result back to the caller.
    """

    def __init__(self, server_type: str, max_tasks: int = 10) -> None:
        self._type      = server_type
        self._max_tasks = max_tasks

    def __call__(self) -> None:
        host = simgrid.this_actor.get_host()
        simgrid.this_actor.info(
            f"{self._type} Server '{host.name}' started "
            f"(max_tasks: {self._max_tasks})"
        )

        mbox = simgrid.Mailbox.by_name(host.name)

        tasks_done = 0
        while self._max_tasks < 0 or tasks_done < self._max_tasks:
            try:
                comm = mbox.get_async()
                comm.wait_for(10.0)          # 10 s idle timeout
                item = comm.get_payload()
            except (simgrid.TimeoutException, simgrid.NetworkFailureException):
                break  # No more tasks arriving – shut down

            if item is None:
                break

            tasks_done += 1
            workload, return_mbox = item
            simgrid.this_actor.info(
                f"[{self._type}] Received task ({workload / 1e6:.2f} MFlops) "
                f"from '{return_mbox}'"
            )

            simgrid.this_actor.execute(workload)

            result = (
                f"Completed {workload / 1e6:.2f} MFlops on "
                f"{self._type} server '{host.name}' at "
                f"t={simgrid.Engine.clock:.2f} s"
            )
            simgrid.this_actor.info(f"[{self._type}] Sending result to '{return_mbox}'")
            simgrid.Mailbox.by_name(return_mbox).put(result, 1_000)

        simgrid.this_actor.info(f"{self._type} Server completed")


def main() -> None:
    e = simgrid.Engine(sys.argv)

    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} <platform_file.xml>", file=sys.stderr)
        sys.exit(1)

    e.load_platform(sys.argv[1])
    hosts = e.all_hosts

    simgrid.this_actor.info("=== Data Offloading Application ===")
    simgrid.this_actor.info(f"Platform loaded with {len(hosts)} hosts")

    edge_hosts  = [h for h in hosts if "edge"  in h.name]
    fog_hosts   = [h for h in hosts if "fog"   in h.name]
    cloud_hosts = [h for h in hosts if "cloud" in h.name]

    # Fallbacks when naming convention is not used
    if not edge_hosts and not fog_hosts and not cloud_hosts:
        if len(hosts) < 3:
            print("Need at least 3 hosts.", file=sys.stderr)
            sys.exit(1)
        edge_hosts  = list(hosts[2:])
        fog_hosts   = [hosts[1]]
        cloud_hosts = [hosts[0]]

    if not fog_hosts or not cloud_hosts:
        print("Platform must have at least one fog and one cloud host.", file=sys.stderr)
        sys.exit(1)

    fog_host   = fog_hosts[0]
    cloud_host = cloud_hosts[0]

    simgrid.this_actor.info(
        f"Servers: fog='{fog_host.name}', cloud='{cloud_host.name}'"
    )
    simgrid.this_actor.info(f"Edge devices: {[h.name for h in edge_hosts]}")

    # Deploy servers
    simgrid.Actor.create("fog_server",   fog_host,   OffloadingServer("FOG",   max_tasks=-1))
    simgrid.Actor.create("cloud_server", cloud_host, OffloadingServer("CLOUD", max_tasks=-1))

    # Deploy edge devices with varying workloads
    workloads = [5e8, 5e9, 5e10]   # low, medium, high
    for i, host in enumerate(edge_hosts):
        wl = workloads[i % len(workloads)]
        simgrid.Actor.create("smart_edge", host,
                             SmartEdgeDevice(fog_host.name, cloud_host.name,
                                             wl, num_tasks=3))

    e.run()
    simgrid.this_actor.info("=== Simulation completed ===")
    simgrid.this_actor.info(f"Simulated time: {simgrid.Engine.clock:.2f} s")


if __name__ == "__main__":
    main()
